tensorflow-core/tensorflow-core-api/src/main/java/org/tensorflow/TensorScope.java

```diff
@@ -0,0 +1,211 @@
+package org.tensorflow;
+
+import java.util.ArrayDeque;
+import java.util.Collections;
+import java.util.Deque;
+import java.util.HashSet;
+import java.util.Iterator;
+import java.util.LinkedHashSet;
+import java.util.Set;
+import java.util.WeakHashMap;
+
+
+/**
+ * A scope that can be used to manage tensor resources.  If auto-attach is used, any tensors created between a scope's
+ * creation and calling {@code close()}, that haven't been detached, are guaranteed to be closed with the scope (even if
+ * they are created in a sub-scope).  Tensors may be manually closed earlier without issue.
+ * <p>
+ * When auto-attach is true, tensors are automatically tracked on creation.  A tensor can me manually added to a scope
+ * with {@link TensorScope#attach(Tensor)} or {@link Tensor#attachToCurrentScope()}.  The tensor will then be closed when the first of it's managing scopes closes.
+ * <p>
+ * {@link Tensor#detach()} detaches the tensor from all scopes, requiring the user to close it manually or attach it to
+ * another scope.
+ * <p>
+ * Note that scope management is thread local, except for detach, which will detach even from scopes on other threads.
+ */
+public class TensorScope implements AutoCloseable {
+
+  private static final Set<TensorScope> allScopes = Collections.newSetFromMap(new WeakHashMap<>());
+
+  private static final ThreadLocal<Deque<TensorScope>> scopeStack = ThreadLocal.withInitial(ArrayDeque::new);
+
+  /**
+   * Attach the tensor to the most recent scope that accepts automatic attachment.
+   *
+   * @return true if attached.
+   */
+  static boolean autoAttach(Tensor tensor) {
+    Iterator<TensorScope> iterator = scopeStack.get().descendingIterator();
+    while (iterator.hasNext()) {
+      TensorScope scope = iterator.next();
+      if (scope.autoAttach) {
+        scope.attach(tensor);
+        return true;
+      }
+    }
+    return false;
+  }
+
+  /**
+   * Return true if there is a scope that accepts auto attachment on the stack.
+   */
+  public static boolean hasAutoScope() {
+    Iterator<TensorScope> iterator = scopeStack.get().descendingIterator();
+    while (iterator.hasNext()) {
+      TensorScope scope = iterator.next();
+      if (scope.autoAttach) {
+        return true;
+      }
+    }
+    return false;
+  }
+
+  /**
+   * Detaches the given tensor from any scopes managing it, requiring it to be manually closed.
+   */
+  public static void detach(Tensor t) {
+    RawTensor raw = t.asRawTensor();
+    synchronized (TensorScope.class) {
+      allScopes.forEach(x -> x.detachTensor(raw));
+    }
+    raw.attached = false;
+  }
+
+  /**
+   * Create a new tensor scope.  If {@code autoAttach} is false, will not automatically manage tensors.
+   *
+   * @see TensorScope
+   */
+  public TensorScope(boolean autoAttach) {
+    this.autoAttach = autoAttach;
+
+    synchronized (TensorScope.class) {
+      allScopes.add(this);
+    }
+
+    localScopeStack = scopeStack.get();
+    localScopeStack.push(this);
```

Comment:
So if this gets stored in a field then any other scopes allocated by that thread get stuck inside it? What happens if someone uses a FJP or other executor service and has a scope in there?

Reply:
See recent changes and comment above.